src/libsyntax/tokenstream.rs

```diff
@@ -206,5 +210,1105 @@ impl TokenTree {
+
+    let mut lo_span = fst_span.lo;
+    let mut hi_span = fst_span.hi;
+
+    // compute the spans iteratively
+    trees.iter()
+        .skip(1)
+        .map(|t| {
+            let sp = t.span();
+            if sp.lo < lo_span {
+                lo_span = sp.lo;
+            }
+            if hi_span < sp.hi {
+                hi_span = sp.hi;
+            }
+        })
+        .count(); // because there isn't a dummy consumer
+                                      // and hopefully this is better than
+                                      // building a vector
+
+    mk_sp(lo_span, hi_span)
+}
+
+/// Convert a vector of `TokenTree`s into a `TokenStream`.
+#[inline(always)]
+pub fn tts_to_ts(trees: Vec<TokenTree>) -> TokenStream {
+    let span = compute_trees_span(&trees);
+    TokenStream {
+        tts: trees,
+        span: span,
+    }
+}
+
+/// Given two `TokenStream`s, return one that contains both
+#[inline]
+pub fn append_streams(ts1: TokenStream, ts2: TokenStream) -> TokenStream {
+    let mut tts1 = ts1.tts;
+    let mut tts2 = ts2.tts;
+    tts1.append(&mut tts2);
+
+    let span = DUMMY_SP; // What _should_ we do here?
+
+    TokenStream {
+        tts: tts1,
+        span: span,
+    }
+}
+
+/// Put a `TokenTree` on the front of a `TokenStream`.
+#[inline]
+pub fn cons_tt_ts(tt: TokenTree, ts: TokenStream) -> TokenStream {
+    let mut tts = Vec::new();
+    tts.push(tt);
+    let mut tts2 = ts.tts;
+    tts.append(&mut tts2);
+
+    let span = DUMMY_SP; // What _should_ we do here?
+
+    TokenStream {
+        tts: tts,
+        span: span,
+    }
+}
+
+/// TokenStream operators include basic destructuring, boolean operations, `maybe_...`
+/// operations, and `maybe_..._prefix` operations. Boolean operations are straightforward,
+/// indicating information about the structure of the stream. The `maybe_...` operations
+/// return `Some<...>` if the tokenstream contains the appropriate item.
+///
+/// Similarly, the `maybe_..._prefix` operations potentially return a
+/// partially-destructured stream as a pair where the first element is the expected item
+/// and the second is the remainder of the stream. As anb example,
+///
+///    `maybe_path_prefix("a::b::c(a,b,c).foo()") -> (a::b::c, "(a,b,c).foo()")`
+impl TokenStream {
+    pub fn respan(self, span: Span) -> TokenStream {
+        TokenStream {
+            tts: self.tts,
+            span: span,
+        }
+    }
+
+    pub fn as_slice(&self) -> &TokenSlice {
+        self
+    }
+
+    pub fn from_ast_lit_str(lit: ast::Lit) -> TokenStream {
+        match lit.node {
+            LitKind::Str(val, _) => {
+                let val = PLit::Str_(token::intern(&val[..]));
+                tts_to_ts(vec![TokenTree::Token(lit.span, Token::Literal(val, None))])
+            }
+            _ => panic!("Invalid literal '{:?}', expected string", lit),
+        }
+
+    }
+
+    pub fn as_paren_delimited_stream(tts: Vec<TokenTree>) -> TokenStream {
```

Comment:
what is this used for? it seems to be at a different level of abstraction to the rest of the functions